src/librustc_mir/transform/instcombine.rs

```diff
@@ -88,7 +88,9 @@ impl Visitor<'tcx> for OptimizationFinder<'b, 'tcx> {
             if let PlaceRef { local, projection: &[ref proj_base @ .., ProjectionElem::Deref] } =
                 place.as_ref()
             {
-                if Place::ty_from(local, proj_base, self.body, self.tcx).ty.is_region_ptr() {
+                // The dereferenced place must have type `&_`.
+                let ty = Place::ty_from(local, proj_base, self.body, self.tcx).ty;
+                if let ty::Ref(_, _, Mutability::Not) = ty.kind {
```

Comment:
wouldn't the right fix be to change the emitted `Rvalue` to use `Operand::Move` for mutable reborrows?

Reply:
Done